### PR TITLE
VM: harden substr/left/right against bad indexes

### LINKED ISSUE #387: VM: check for the slice bounds explicitly
There is a case in #384 where previous version of Go failed some of the testcases.
Funnily enough, we can make the same tests fail on the current versions of Go just by changing
```golang
vm.estack.PushVal([]byte("abcdef"))
```
to
```golang
s := "abcdef"
vm.estack.PushVal([]byte(s))
```

This is something we should fix, by explicitly checking for slice bounds where possible.
This has an additional benefit of more explicit errors.

### PR DESCRIPTION
**Pull request by roman-khimov**
And improve PUSHDATA along the way. Fixes #387.

### CODE REVIEW

**Comment by roman-khimov** [comment]
Fixed gofmt.

File: pkg/vm/vm_test.go [hunk 1]
```diff
@@ -925,6 +925,15 @@ func TestLEFTBadNoString(t *testing.T) {
 	assert.Equal(t, true, vm.state.HasFlag(faultState))
 }
 
+func TestLEFTBadNegativeLen(t *testing.T) {
```
Comment (volekerb) [inline pkg/vm/vm_test.go]: nit: what about having tests with params with testdata instead of copy pasting same code for tests?

Comment (roman-khimov) [reply pkg/vm/vm_test.go]: Dunno. At the moment all of these tests are trivial and even though they have some things in common, that's on the order of 3-5 lines which is not a lot to win by making them more complex. And I still hope that #196 will solve most of our VM testing problems.

File: pkg/vm/vm.go [hunk 2]
```diff
@@ -311,16 +320,31 @@ func (v *VM) execute(ctx *Context, op Instruction) {
 		l := int(v.estack.Pop().BigInt().Int64())
 		o := int(v.estack.Pop().BigInt().Int64())
 		s := v.estack.Pop().Bytes()
+		if l < 0 {
+			panic("negative index")
```
Comment (fyrchik) [inline pkg/vm/vm.go]: shouldn't it be the other way? (here `negative length`, 3 lines below -- `negative index`)

Comment (roman-khimov) [reply pkg/vm/vm.go]: Yep, thanks. Will fix.

File: pkg/vm/vm.go [hunk 3]
```diff
@@ -311,16 +320,31 @@ func (v *VM) execute(ctx *Context, op Instruction) {
...
+		}
+		if l+o > len(s) {
+			panic("out of bounds access")
+		}
 		v.estack.PushVal(s[o : o+l])
 	case LEFT:
 		l := int(v.estack.Pop().BigInt().Int64())
+		if l < 0 {
+			panic("negative index")
```
Comment (fyrchik) [inline pkg/vm/vm.go]: Here and in `RIGHT` it is also `length` I believe.

Comment (roman-khimov) [reply pkg/vm/vm.go]: Fixed.

**Pull request merged**